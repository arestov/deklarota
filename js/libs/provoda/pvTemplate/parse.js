
import getCachedPVData from './getCachedPVData'

function getCommentPVData(cur_node, struc_store, getSample) {
  return getCachedPVData(cur_node, struc_store, true, getSample)
}

function getPVData(cur_node, struc_store, getSample) {
  return getCachedPVData(cur_node, struc_store, false, getSample)
}

export default function parse(start_node, struc_store, getSample, _opts) {
  //полный парсинг, без байндинга

  var result = []

  var match_stack = [ start_node ]
  var i = 0
  while (match_stack.length) {
    var cur_node = match_stack.shift()
    var node_type = cur_node.nodeType
    var directives_data = null
    if (node_type == 1) {
      directives_data = getPVData(cur_node, struc_store, getSample)
      result.push(cur_node, directives_data)
    } else if (node_type == 8) {
      directives_data = getCommentPVData(cur_node, struc_store, getSample)
      result.push(cur_node, directives_data)
    }

    // if (directives_data.replacer) {
    // 	match_stack.unshift(directives_data.node);
    // }

    if (node_type == 1) {
      for (i = 0; i < cur_node.childNodes.length; i++) {
        match_stack.push(cur_node.childNodes[i])
      }
    }

  }
  return result
}
