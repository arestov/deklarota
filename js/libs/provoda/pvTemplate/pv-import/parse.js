

export default function parsePVImport(node, sample_name) {

  var possible = node.nodeName === 'SCRIPT'
    ? node
    : node.querySelector('script[type="pv-import-map"]')

  var script
  if (possible === node) {
    script = node
  }

  if (possible.parentNode === node) {
    script = node
    node.removeChild(script)
  }
  var map_string = script && script.textContent
  var map = map_string ? JSON.parse(map_string) : [{}]

  return {
    sample_name: sample_name,
    map: map,
    pv_nest: script.getAttribute('pv-nest') || null
  }
};
