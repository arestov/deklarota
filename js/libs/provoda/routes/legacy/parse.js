
import spv from '../../../spv'

// from template to full string - implemented
// from string to match - NOT IMPLEMENTED

// tracks/[:trackName]/more/something/else
// how to match subpaths? submodels can route same path
// subpaths levels handling should be limited to 2 levels for now
// so model can handle only /tracks or /tracks/[:trackName] but not tracks/[:trackName]/more
// example "#tracks/[:artist:next_value],[:track]" // artist:next_value - way to map data
// {next_value: 'Mike'} will be used as "artist" for template

const string_state_regexp = /\[\:.+?\]/gi

const makeStatesMap = function(states_raw) {
  const result = new Array(states_raw.length)

  for (let i = 0; i < states_raw.length; i++) {
    const parts = states_raw[i].split(':')
    const dest = parts[0]
    const source = parts[1] || dest
    const value = parts[2]
    result[i] = [dest, source, value]
  }

  return result
}

const parse = function(full_usable_string) {
  // tracks/[:artist],[:track]
  const clean_string_parts = full_usable_string.split(string_state_regexp)
  const states = full_usable_string.match(string_state_regexp)

  if (states) {
    for (let i = 0; i < states.length; i++) {
      states[i] = states[i].slice(2, states[i].length - 1)
    }
  }

  const states_map = states && makeStatesMap(states)

  return {
    clean_string_parts: clean_string_parts,
    states: states,
    states_map: states_map,
  }
}

const parsePath = spv.memorize(parse)

export default parsePath
