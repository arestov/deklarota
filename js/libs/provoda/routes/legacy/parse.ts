
import memorize from '../../../spv/memorize'

// from template to full string - implemented
// from string to match - NOT IMPLEMENTED

// tracks/[:trackName]/more/something/else
// how to match subpaths? submodels can route same path
// subpaths levels handling should be limited to 2 levels for now
// so model can handle only /tracks or /tracks/[:trackName] but not tracks/[:trackName]/more
// example "#tracks/[:artist:next_value],[:track]" // artist:next_value - way to map data
// {next_value: 'Mike'} will be used as "artist" for template

const string_state_regexp = /\[\:.+?\]/gi

type LegacyRouteParamsMap = [string, string, string | undefined][]

const makeStatesMap = function(states_raw: string[]):LegacyRouteParamsMap {
  const result: LegacyRouteParamsMap = new Array(states_raw.length)

  for (let i = 0; i < states_raw.length; i++) {
    const parts = (states_raw[i] as string).split(':')
    const dest = parts[0]
    if (dest == null) {
      throw new Error()
    }
    const source = parts[1] || dest
    const value = parts[2]
    result[i] = [dest, source, value]
  }

  return result
}

type ParsedLegacyRoute = {
  clean_string_parts: string[]
  states: string[] | null
  states_map: LegacyRouteParamsMap | null
}

const parse = function(full_usable_string: string): ParsedLegacyRoute {
  // tracks/[:artist],[:track]
  const clean_string_parts = full_usable_string.split(string_state_regexp)
  const states = full_usable_string.match(string_state_regexp)

  if (states) {
    for (let i = 0; i < states.length; i++) {
      const cur = states[i] as string
      states[i] = cur.slice(2, cur.length - 1)
    }
  }

  const statesconv = states as string[] | null

  const states_map = states && makeStatesMap(states)

  return {
    clean_string_parts: clean_string_parts,
    states: statesconv,
    states_map: states_map,
  }
}

const parsePath = memorize(parse)

export default parsePath
