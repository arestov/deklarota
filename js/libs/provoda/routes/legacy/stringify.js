define(function(require) {
'use strict'

var getValue = function(app, value) {
  if (value) {
    return app.encodeURLPart(value)
  }

  if (value === 0) {
    return app.encodeURLPart(value)
  }
}

var pathExecutor = function(getChunk) {
  return function getPath(obj, app, arg1, arg2) {
    if (obj.parsed.states) {
      var full_path = ''
      for (var i = 0; i < obj.parsed.clean_string_parts.length; i++) {
        full_path += obj.parsed.clean_string_parts[i]
        var cur_state = obj.parsed.states[i]
        if (cur_state) {
          var chunk = getChunk(cur_state, app, arg1, arg2)
          full_path += getValue(app, chunk) || 'null'
        }
      }
      return full_path
    }
    return obj.full_usable_string
  }
}

return pathExecutor

})
