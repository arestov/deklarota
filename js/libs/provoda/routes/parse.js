
// var spv = require('spv')

const string_state_regexp = /((.*?)(\[\:(.+?)\]))|(.+)/g
// 0:                     1:((2:)(3:(4:)))|(:5)
const PREFIX = 2
const STATE_SOURCE = 3
const STATE = 4
const ALTERNATIVE = 5

// Add optional params - 'tracks/[:artist:next_value],([:track])'
// Add optional values - 'tracks/[:artist::Smith],([:track])'

const stateItem = function(text) {
  const parts = text.split(':')
  const dest = parts[0]
  const source = parts[1] || dest
  const value = parts[2]
  if (value == '') {
    throw new Error('value can\'t be empty string')
  }
  if (value && parts[1]) {
    throw new Error('param should be empty when value exists')
  }
  return [dest, source, value]
}

const reRegExpChar = /[\\^$.*+?()[\]{}|]/g
const reHasRegExpChar = RegExp(reRegExpChar.source)


const escapeRegExp = function(string) {
  return (string && reHasRegExpChar.test(string))
    ? string.replace(reRegExpChar, '\\$&')
    : string
}

const createMatcher = function(prefix, rest) {
  if (!prefix && !rest) {
    throw new Error('should be prefix or rest')
  }

  if (!rest) {
    return {
      regexp: new RegExp(escapeRegExp(prefix)),
      has_group: false,
    }
  }

  const paramsRegExp = '(.+?)'

  if (!prefix) {
    return {
      regexp: new RegExp(paramsRegExp),
      has_group: true,
    }
  }

  return {
    regexp: new RegExp(escapeRegExp(prefix) + paramsRegExp),
    has_group: true,
  }

}

const parse = function(full_usable_string) {
  // tracks/[:artist],[:track]
  const iterableCopy = new RegExp(string_state_regexp.source, string_state_regexp.flags)
  let cur_matches
  const parts = []
  let cur_matching_group = 1 // 0 is whole match
  let prev_have_state = false
  do {
    cur_matches = iterableCopy.exec(full_usable_string)

    if (!cur_matches) {
      break
    }

    const prefix = cur_matches[PREFIX]
    const state_source = cur_matches[STATE_SOURCE]

    if (!state_source) {
      const alternative = cur_matches[ALTERNATIVE]
      parts.push({
        prefix: alternative,
        state: null,
        state_source: null,
        matching_group: null,
        matcher: createMatcher(alternative, false)
      })
      prev_have_state = false
      continue
    }

    const state_dcl = cur_matches[STATE]
    const matching_group = cur_matching_group

    if (!prefix && prev_have_state) {
      const prev = parts[parts.length - 1]
      throw new Error(
        'Should be any separator between ' +
        prev.state_source + ' and ' + state_source + '.' +
        'Use comma, space, minus, plus, underscore'
      )
    }

    parts.push({
      prefix: prefix,
      state: stateItem(state_dcl),
      state_source: state_source,
      matching_group: matching_group,
      matcher: createMatcher(prefix, true),
    })
    cur_matching_group = cur_matching_group + 1
    prev_have_state = true
  } while (cur_matches)

  // TODO: check if two parts with state have separator
  // TODO: check to have only uniq params

  let fullRegExpString = '^'
  for (let i = 0; i < parts.length; i++) {
    fullRegExpString += parts[i].matcher.regexp.source
  }

  fullRegExpString += '$'

  return {
    parts: parts,
    matcher: new RegExp(fullRegExpString),
  }
}

export default parse
