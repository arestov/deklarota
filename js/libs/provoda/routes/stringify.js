
import looksLikeWrappedJSON from './utils/looksLikeWrappedJSON'

var statePartAsString = function(state_dcl, data) {
  if (!state_dcl) {
    return ''
  }

  var source = state_dcl[1]
  var value = data[source]
  if (value == null) {
    return ''
  }

  if (typeof value == 'string') {
    if (looksLikeWrappedJSON(value)) {
      return encodeURIComponent(JSON.stringify(data[source]))
    }

    return encodeURIComponent(value)
  }

  return encodeURIComponent('"' + JSON.stringify(data[source]) + '"')
}


var toString = function(parsed, data) {
  // 'tracks/[:artist:next_value],[:track]' +  {next_value: 'Mike', track: 'Play With You'}
  // => 'tracks/Mike,Play%20With%20You'

  var parts = parsed.parts
  var toStrings = new Array(parts.length)
  for (var i = 0; i < parts.length; i++) {
    var cur = parts[i]
    var result = (cur.prefix || '') + statePartAsString(cur.state, data)
    toStrings[i] = result
  }
  return toStrings.join('')
}

export default toString
