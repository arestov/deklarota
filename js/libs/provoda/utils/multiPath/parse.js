

import spv from '../../../spv'
import getParsedPath from '../../routes/legacy/getParsedPath'
import supportedZip from './supportedZip'
import fromLegacy from './fromLegacy'
import isJustAttrAddr from './isJustAttrAddr'

var splitByDot = spv.splitByDot
var empty = Object.freeze({})
var root = Object.freeze({
  type: 'root',
  steps: null,
})
var parents = spv.memorize(function(num) {
  return Object.freeze({
    type: 'parent',
    steps: num,
  })
})
var parent_count_regexp = /\^+/gi

/*


/(\^|\s+)(\<)(\s+)/
"< @all state_name < nesting < resource < #"

"< state_name < aggr:nesting < resource < #"

"< state_name << /resource/[:ddaf]/sdf < #"

"< state_name <<< #"
"<< nesting_name << #"

"<< nesting_name << ^^"
"< state_name <<< ^^"

"< state_name"
"state_name"

"/resource/[:ddaf]/sdf < #"
"/resource/[:ddaf]/sdf <"

"nesting_name < < ^^"
*/
var checkSplit = /(?:^|\s+)?<(?:\s+)?/
var end = /(<$)|(\^$)|(#$)/
var start = /^</

var parseFromStart = spv.memorize(function(string) {
  var parts = string.split(checkSplit)
  // parts[0] should be empty
  var state = parts[1]
  var nest = parts[2]
  var resource = parts[3]
  var base = parts[4]

  return parseParts(state, nest, resource, base)

})

var parseFromEnd = spv.memorize(function(string) {
  var parts = string.split(checkSplit)

  var length = parts.length
  var base = parts[length - 1]
  var resource = parts[length - 2]
  var nest = parts[length - 3]
  var state = parts[length - 4]

  return parseParts(state, nest, resource, base)
})

function canParseModern(string) {
  var from_start = start.test(string)
  var from_end = end.test(string)
  return (from_start || from_end)
    ? {from_start: from_start, from_end: from_end}
    : null
}

var parseModern = spv.memorize(function parseModern(string) {
  var can_parse = canParseModern(string)
  if (can_parse == null) {
    return null
  }

  if (can_parse.from_start) {
    return parseFromStart(string)
  }
  return parseFromEnd(string)
})

var matchNotStateSymbols = /(^\W)|\@|\:/


export var getStateInfo = spv.memorize(function getStateInfo(string) {
  if (!string) {
    return empty
  }

  return {
    base: splitByDot(string)[0],
    path: string,
  }
})

var SimpleStateAddr = function(string) {
  this.state = getStateInfo(string)
}

SimpleStateAddr.prototype = spv.cloneObj(SimpleStateAddr.prototype, {
  result_type: 'state',
  zip_name: null,
  state: null,
  nesting: empty,
  resource: empty,
  from_base: empty,
  as_string: null,
})

var simpleState = spv.memorize(function simpleState(string) {
  return new SimpleStateAddr(string)
})

var attemptSimpleStateName = function(string) {
  if (!string || matchNotStateSymbols.test(string)) {
    return null
  }

  return simpleState(string)
}

var self = {
  as_string: '<<<<',
  base_itself: true,
}

var parseMultiPath = function(string, allow_legacy) {
  if (string == '<<<<') {
    return self
  }


  var modern = parseModern(string)
  if (modern != null) {
    var state_info = modern.state
    if (state_info && state_info.base && !isJustAttrAddr(modern) && state_info.base.startsWith('__')) {
      console.warn(new Error('private attr used'), string)
    }

    return modern
  }

  return attemptSimpleStateName(string) || (
    allow_legacy ? fromLegacy(string) : null
  )

}
var matchZip = /(?:\@(.+?)\:)?(.+)?/

var parseLegacyCached = spv.memorize(parseMultiPath)
var parseModernCached = spv.memorize(parseMultiPath)


var parseWithCache = function(addr_str, legacy_ok) {
  if (legacy_ok != true) {
    return parseModernCached(addr_str, false)
  }

  return parseLegacyCached(addr_str, true)
}


parseWithCache.simpleState = simpleState

export default parseWithCache

function parseParts(state_raw, nest_raw, resource_raw, base_raw) {
  var state_part_splited = state_raw && state_raw.match(matchZip)
  var zip_state_string = state_part_splited && state_part_splited[1]
  var state_string = state_part_splited && state_part_splited[2]

  var nest_part_splited = nest_raw && nest_raw.match(matchZip)
  var zip_nest_string = nest_part_splited && nest_part_splited[1]
  var nest_string = nest_part_splited && nest_part_splited[2]

  if (zip_state_string && zip_nest_string) {
    throw new Error('both state and nesting cant have zip_name')
  }

  if (zip_nest_string && state_string) {
    throw new Error('use state zip')
  }

  if (zip_state_string && !state_string && nest_string) {
    throw new Error('use nest zip')
  }

  if (zip_state_string && !supportedZip(zip_state_string, 'state')) {
    throw new Error('unsupported zip for state: ' + zip_state_string)
  }

  if (zip_nest_string && !supportedZip(zip_nest_string, 'nesting')) {
    throw new Error('unsupported zip for state: ' + zip_nest_string)
  }


  var zip_name = zip_state_string || zip_nest_string || null
  var state_info = getStateInfo(state_string)
  var nest_info = getNestInfo(nest_string)
  var resource_info = getResourceInfo(resource_raw)
  var base_info = getBaseInfo(base_raw)

  var result_type = getResultType(state_info, nest_info, resource_info, base_info)

  return {
    result_type: result_type,
    zip_name: zip_name,
    state: state_info,
    nesting: nest_info,
    resource: resource_info,
    from_base: base_info,
    as_string: null,
  }
}


export function updateResultType(draft) {
  draft.result_type = getResultType(draft.state, draft.nesting, draft.resource, draft.from_base)

}

export function createAddrByPart(source) {
  var draft = {
    result_type: null,
    zip_name: null,
    state: source.state || empty ,
    nesting: source.nesting || empty,
    resource: source.resource || empty,
    from_base: source.from_base || empty,
    as_string: null,
  }

  updateResultType(draft)
  return draft
}


export function getNestInfo(string) {
  if (!string) {
    return empty
  }

  var parts = string.split(':')
  var path = parts.pop()

  var full_path = splitByDot(path)

  var zip_name = parts[0] || null

  if (zip_name) {
    throw new Error('dont use. use < @[zip_name] [statename] < [nestingname]')
  }

  var target_nest_name = full_path[full_path.length - 1] // last one

  if (!target_nest_name) {
    throw new Error('wrong nest path: ' + string)
  }

  return {
    path: full_path,
    base: full_path.slice(0, full_path.length - 1), // all, except last
    target_nest_name: target_nest_name,
  }
}

export function getResourceInfo(string) {
  if (!string) {
    return empty
  }

  if (string.startsWith('#')) {
    throw new Error('use "ascending part" for root/parent traversing')
  }

  return {
    path: string,
    template: getParsedPath(string),
  }
}

export function getBaseInfo(string) {
  if (!string) {
    return empty
  }

  if (string == '#') {
    return root
  }

  var from_parent_num = string.match(parent_count_regexp)
  if (from_parent_num) {
    return parents(from_parent_num[0].length)
  }

  throw new Error('unsupported base: ' + string)
}

function getResultType(state, nest) {
  if (state && state.path) {
    return 'state'
  }

  if (nest && nest.path) {
    return 'nesting'
  }

  return null
}
