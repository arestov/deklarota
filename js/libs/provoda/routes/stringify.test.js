import {describe, it, expect} from '@jest/globals'

import parse from './parse'
import stringify from './stringify'

describe('parse', () => {
  describe('smoke', () => {
    it.each([
      ['tracks/[:artist]', {artist: 'Smith'}],
      ['tracks/[:artist],[:track]', {artist: 'Smith', track: 'nice to have you'}],
      ['tracks/[:artist:next_value],([:track])', {next_value: 'Smith', track: 'nice to have you'}],
      ['tracks/[:artist::Smith]', {artist: 'Smith'}],
    ])('result should match snapshot for "%s"', (template, data) => {
      expect(stringify(parse(template), data)).toMatchSnapshot()
    })
  })
})
