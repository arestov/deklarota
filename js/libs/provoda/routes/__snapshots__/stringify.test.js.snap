// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse smoke result should match snapshot for "tracks/[:artist::Smith]" 1`] = `"tracks/Smith"`;

exports[`parse smoke result should match snapshot for "tracks/[:artist:next_value],([:track])" 1`] = `"tracks/Smith,(nice%20to%20have%20you)"`;

exports[`parse smoke result should match snapshot for "tracks/[:artist]" 1`] = `"tracks/Smith"`;

exports[`parse smoke result should match snapshot for "tracks/[:artist],[:track]" 1`] = `"tracks/Smith,nice%20to%20have%20you"`;
